Flav/int to bigint migration

flvndvd opened the pull request:
<!-- Briefly describe the changes you've made and link any relevant issues (e.g., "Fixes #123"). -->
<!-- If the PR includes UI changes, please attach a screenshot or GIF to illustrate the modifications. -->
# Integer to BigInt Migration Tool

See migration details [here](https://www.notion.so/dust-tt/Migrating-Primary-Keys-from-INT-to-BIGINT-in-PostgreSQL-14b28599d94180a4b280c17f7f99e453?d=14b28599d94180a38006001ccf448dae#14b28599d94180faa780d48a0c3cbbcd).

This PR introduces a PostgreSQL migration tool designed to safely convert `integer` primary/foreign keys to `BigInt` while maintaining zero-downtime capability and full rollback support.

## Key Features

- **Zero-downtime migration**: Uses concurrent index creation and maintains dual columns during transition
- **Phased approach**: Splits migration into distinct steps (setup, backfill, pre-cutover, cutover, rollback)
- **Data integrity**: Maintains referential integrity throughout the process using triggers
- **Rollback support**: Full rollback capability at any stage
- **Safety measures**: 
  - Dry-run mode for validation
  - Progress tracking
  - Transaction safety
  - Concurrent index creation
  - NOT VALID + VALIDATE pattern for constraints

## Key Implementation Details

### Constraint Handling

1. **Standardized Constraint Behavior**
   - All foreign key constraints are enforced to use `ON UPDATE RESTRICT ON DELETE RESTRICT`
   - This applies to both new constraints and rollback scenarios
   - Even when rolling back, we don't restore the original CASCADE behaviors
   - This is an intentional design decision to maintain consistent referential integrity rules

2. **Primary Key Dependencies**
   - PostgreSQL requires dropping foreign key constraints before modifying primary key constraints
   - This is because foreign keys create dependencies that prevent primary key modifications
   - The original foreign keys will get restored when using the `rollback` step 
   
### Rollback Considerations

1. **Constraint Behavior Changes**
   - Original: Mixed ON UPDATE/DELETE behaviors (CASCADE/RESTRICT)
   - After rollback: Standardized to RESTRICT for both
   - This is the only permanent change after rollback
   - Assuming we properly delete each resources individually and don't rely on the `CASCADE` effect. 

2. **Data Integrity During Rollback**
   - Maintains both integer and `BigInt` columns
   - Preserves data synchronization via triggers
   - Ensures zero data loss during rollback

## Migration Steps

### 1. Setup
- Adds new `BigInt` columns (`_new` suffix)
- Creates sync triggers to maintain data consistency (`id` -> `id_new`)
- Preserves existing constraints

### 2. Backfill
- Copies data from `integer` to `BigInt` columns
- Batched processing to minimize impact
- Progress tracking with percentage completion

### 3. Pre-cutover
- Creates necessary indexes concurrently
- Sets up constraints for new columns
- Validates data consistency
- Don't require any transaction

### 4. Cutover
- Switches primary key to `BigInt` column
- Updates foreign key references
- Maintains rollback capability via legacy columns

### 5. Rollback (if needed)
- Restores integer primary keys
- Rebuilds original foreign key relationships
- Maintains data sync for safety

## Usage Example

```typescript
// Setup phase
npx tsx migration.ts --database=postgresql://... --table=users --step=setup

// Backfill phase
npx tsx migration.ts --database=postgresql://... --table=users --step=backfill --batchSize=5000

// Pre-cutover phase
npx tsx migration.ts --database=postgresql://... --table=users --step=pre-cutover

// Cutover phase
npx tsx migration.ts --database=postgresql://... --table=users --step=cutover

// Rollback (if needed)
npx tsx migration.ts --database=postgresql://... --table=users --step=rollback


## Risk

<!-- Discuss potential risks and how they will be mitigated. Consider the impact and whether the changes are safe to rollback. -->

## Deploy Plan

<!-- Outline the deployment steps. Specify the order of operations and any considerations that should be made before, during, and after deployment/ -->

flvndvd committed:
Add migration to update primary/foreign keys from int to BigInt.

flvndvd committed:
Better structure steps in script

flvndvd committed:
Tmp

flvndvd committed:
Add cutover/rollback migration

flvndvd committed:
✨

flvndvd committed:
✂️

flvndvd commented on front/lib/resources/storage/index.ts:
```diff
@@ -8,6 +10,19 @@ const acquireAttempts = new WeakMap();
+// prevents silent precision loss when handling large integers from the database.
+// Throws an assertion error if a BIGINT value exceeds JavaScript's safe integer
+// limits.
+types.setTypeParser(builtins.INT8, function (val) {
```
Will extract to a dedicated PR.

flvndvd replied:
Shipped [here](https://github.com/dust-tt/dust/pull/9067).

spolu commented on front/migrations/20241129_migrate_int_to_bigint.ts:
```diff
@@ -0,0 +1,1438 @@
+    return progress;
+  }
+
+  private async checkTableProgress(
```
Does this fly on a very big table?

spolu commented on front/migrations/20241129_migrate_int_to_bigint.ts:
```diff
@@ -0,0 +1,1438 @@
+    }
+  }
+
+  private async dropTriggers(
```
nit: Should be moved below the main functions

spolu reviewed (approved):
LGTM. Very nice \o/

I presume final clean-up is still missing.

Let's start with connectors :+1:

JulesBelveze commented on front/migrations/20241129_migrate_int_to_bigint.ts:
```diff
@@ -0,0 +1,1438 @@
+        currentColumn: MAIN_ID_COLUMN,
+        newColumn: createColumnName.new(MAIN_ID_COLUMN),
+        legacyColumn: createColumnName.legacy(MAIN_ID_COLUMN),
+        sequenceName: `${this.config.tableName}_id_seq`,
```
```suggestion
        sequenceName: createSequenceName(this.config.tableName),
```

flvndvd replied:
Good catch!

JulesBelveze commented on front/migrations/20241129_migrate_int_to_bigint.ts:
```diff
@@ -0,0 +1,1438 @@
+          currentColumn: MAIN_ID_COLUMN,
+          newColumn: createColumnName.new(MAIN_ID_COLUMN),
+          legacyColumn: createColumnName.legacy(MAIN_ID_COLUMN),
+          sequenceName: `${this.config.tableName}_id_seq`,
```
```suggestion
          sequenceName: createSequenceName(this.config.tableName),
```

JulesBelveze reviewed (commented):
Looks very solid 🔥

flvndvd committed:
Add clean up step

flvndvd commented on front/migrations/20241129_migrate_int_to_bigint.ts:
```diff
@@ -0,0 +1,1438 @@
+export const createTriggerNames = {
+  // For primary key (main table)
+  pk: (tableName: string, direction: SyncDirection) => ({
+    trigger: `${tableName}_pk_sync_${direction}` as const,
+    function: `${tableName}_pk_sync_${direction}_trigger` as const,
```
It's to surface that the function is only used by a trigger.

flvndvd committed:
Address comments from review

flvndvd committed:
Fix non nullable progress

flvndvd committed:
Handle whereClause on composite index

flvndvd committed:
✨

flvndvd committed:
✂️

flvndvd committed:
Release

JulesBelveze committed:
[front/migrations] - feature: enhance foreign key references with actions for update and delete

 - Added fields to track delete and update actions on foreign key constraints in migration script
 - The migration script now includes additional SQL JOIN to fetch action information from the pg_constraint table

JulesBelveze committed:
[front/migrations] - refactor: update foreign key actions in bigint migration

 - Change foreign key `ON UPDATE` and `ON DELETE` actions to dynamic based on reference settings
 - Set actions to `SET NULL` when reference actions are marked with "r", otherwise keep as `RESTRICT`

JulesBelveze committed:
[front] - fix: correct ON UPDATE/DELETE actions in foreign key constraints

 - Fixed an issue where the update and delete actions for foreign key constraints were incorrectly set to RESTRICT when they should be SET NULL
 - Updated the conditional logic to match 'n' values to SET NULL action instead of 'r' in the migration script

flvndvd committed:
Add support for minId

flvndvd committed:
Account for delta

flvndvd committed:
Assert index name length

flvndvd committed:
Shorten index name

flvndvd committed:
Shorten table name

flvndvd committed:
Only assert in production

flvndvd committed:
Merge branch 'main' into flav/int-to-bigint-migration

flvndvd committed:
✨

flvndvd committed:
Add dev types

fontanierh reviewed (approved):
Works great
